src/greenlight_alby_client.rs

```diff
@@ -0,0 +1,217 @@
+use gl_client::signer::Signer;
+use gl_client::tls::TlsConfig;
+
+#[derive(Error, Clone, Debug)]
+pub enum SdkError {
+    #[error("invalid argument: {0}")]
+    InvalidArgument(String),
+
+    #[error("greenlight API error: {0}")]
+    GreenlightApi(String),
+
+    #[error("other error: {0}")]
+    Other(String),
+}
+
+impl SdkError {
+    fn invalid_arg(e: anyhow::Error) -> Self {
+        SdkError::InvalidArgument(Self::format_anyhow_error(e))
+    }
+
+    fn greenlight_api(e: anyhow::Error) -> Self {
+        SdkError::GreenlightApi(Self::format_anyhow_error(e))
+    }
+
+    fn other(e: anyhow::Error) -> Self {
+        SdkError::Other(Self::format_anyhow_error(e))
+    }
+
+    fn format_anyhow_error(e: anyhow::Error) -> String {
+        // Use alternate format (:#) to get the full error chain.
+        format!("{:#}", e)
+    }
+}
+
+pub type Result<T> = std::result::Result<T, SdkError>;
+
+#[derive(Clone, Debug)]
+pub struct GreenlightCredentials {
+    pub device_cert: String,
+    pub device_key: String,
+}
+
+impl From<scheduler::RecoveryResponse> for GreenlightCredentials {
+    fn from(recovery: scheduler::RecoveryResponse) -> Self {
+        GreenlightCredentials {
+            device_cert: recovery.device_cert,
+            device_key: recovery.device_key,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct GreenlightNodeInfo {
+    pub alias: String,
+    pub color: String,
+    pub network: String,
+    pub block_height: u32,
+}
+
+impl From<cln::GetinfoResponse> for GreenlightNodeInfo {
+    fn from(info: cln::GetinfoResponse) -> Self {
+        GreenlightNodeInfo {
+            alias: info.alias.unwrap_or_default(),
+            color: hex::encode(info.color),
+            network: info.network,
+            block_height: info.blockheight,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct GreenlightInvoiceRequest {
+    pub amount_msat: u64,
+    pub description: String,
+    pub label: String,
+}
+
+impl From<GreenlightInvoiceRequest> for cln::InvoiceRequest {
+    fn from(req: GreenlightInvoiceRequest) -> Self {
+        cln::InvoiceRequest {
+            label: req.label,
+            amount_msat: Some(AmountOrAny {
+                value: Some(Value::Amount(Amount {
+                    msat: req.amount_msat,
+                })),
+            }),
+            description: req.description,
+            ..Default::default()
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct GreenlightInvoiceResponse {
+    pub bolt11: String,
+}
+
+impl From<cln::InvoiceResponse> for GreenlightInvoiceResponse {
+    fn from(invoice: cln::InvoiceResponse) -> Self {
+        GreenlightInvoiceResponse {
+            bolt11: invoice.bolt11,
+        }
+    }
+}
+
+pub struct GreenlightAlbyClient {
+    // signer: gl_client::signer::Signer,
+    scheduler: gl_client::scheduler::Scheduler,
+    tls: TlsConfig,
+}
+
+pub async fn recover(mnemonic: String) -> Result<GreenlightCredentials> {
```

Comment:
Can we make this return a pointer? so in go the value would be nil if there is an error